src/libqof/qof/qofsession.cpp

```diff
@@ -124,528 +81,364 @@ qof_backend_get_registered_access_method_list(void)
-    qof_session_init(session);
-    return session;
+    return new QofSessionImpl;
 }
 
-QofBook *
-qof_session_get_book (const QofSession *session)
+void
+QofSessionImpl::destroy_backend () noexcept
 {
-    if (!session) return NULL;
-    if (!session->book) return NULL;
-
-    if ('y' == session->book->book_open)
+    if (m_backend)
     {
-        return session->book;
-    }
-    else
-    {
-        return NULL;
+        clear_error ();
+        if (m_backend->destroy_backend)
+            m_backend->destroy_backend (m_backend);
+        else
+            g_free(m_backend);
+        m_backend = nullptr;
     }
 }
 
-QofBackend *
-qof_session_get_backend (const QofSession *session)
-{
-    if (!session) return NULL;
-    return session->backend;
-}
-
-const char *
-qof_session_get_file_path (const QofSession *session)
-{
-    if (!session) return NULL;
-    if (!session->backend) return NULL;
-    return session->backend->fullpath;
-}
-
-const char *
-qof_session_get_url (const QofSession *session)
-{
-    if (!session) return NULL;
-    return session->book_id;
-}
-
 void
-qof_session_ensure_all_data_loaded (QofSession *session)
-{
-    QofBackend* backend;
-
-    if (session == NULL) return;
-    backend = qof_session_get_backend(session);
-    if (backend == NULL) return;
-
-    if (backend->load == NULL) return;
-    backend->load(backend, qof_session_get_book(session), LOAD_TYPE_LOAD_ALL);
-    qof_session_push_error (session, qof_backend_get_error(backend), NULL);
-}
-
-/* ====================================================================== */
-
-/** Programs that use their own backends also need to call
-the default QOF ones. The backends specified here are
-loaded only by applications that do not have their own. */
-struct backend_providers
-{
-    const char *libdir;
-    const char *filename;
-};
-
-static void
-qof_session_load_backend(QofSession * session, const char * access_method)
+QofSessionImpl::load_backend (std::string access_method) noexcept
 {
-    GSList *p;
-    QofBackendProvider *prov;
-    char *msg;
-    gboolean prov_type;
-    gboolean (*type_check) (const char*);
-
-    ENTER (" list=%d, initted=%s", g_slist_length(provider_list),
-           qof_providers_initialized ? "true" : "false");
-    prov_type = FALSE;
-    if (!qof_providers_initialized)
-    {
-        qof_providers_initialized = TRUE;
-    }
-    p = provider_list;
-    while (p != NULL)
+    ENTER (" list=%lu", provider_list.size ());
+    bool prov_type {false};
+    for (auto const & prov : provider_list)
     {
-        prov = static_cast<QofBackendProvider*>(p->data);
         /* Does this provider handle the desired access method? */
-        if (0 == g_ascii_strcasecmp (access_method, prov->access_method))
+        if (!boost::iequals (access_method, prov->access_method))
+            continue;
+        /* More than one backend could provide this
+        access method, check file type compatibility. */
```

Comment:
That very logic is found on the very next line :-). I took out this `access_method` check, and the `backend/dbi/test/test-backend-dbi-basic.cpp` test fails because the mysql provider doesn't define `check_data_type`: it relies on `access_method` which is `mysql` for the mysql provider, and `file` for an `.xml` file (an `.xml` file is used in this unit test).

I'm thinking it is out of scope to implement a `check_data_type` for the mysql provider at this time, but I agree that string comparisons don't make sense either. As you said, what makes sense would be a `enum class access_method` enumerating the supported access methods.

Do you have another suggestion to get rid of the string comparison, or shall we bide its existence longer?